OSDOCS#6017 cert-manager : Configuring an ACME issuer (#59702)
openshift/openshift-docs: modules/cert-manager-acme-dns01-explicit-gcp.adoc

Reviewer: For this file, also suggest formatting the user-replaceable values with `<brackets_and_underscores>`

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-explicit-gcp_{context}"]

> = Configuring an ACME issuer to solve DNS-01 challenges by using explicit credentials on GCP

Reviewer 2: ```suggestion
= Configuring an ACME issuer to solve DNS-01 challenges by using explicit credentials for GCP Cloud DNS
```

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers-only'       <2>
----
<1> Add the `spec.controllerConfig` section.
<2> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.

.. Save the file to apply the changes.

. Optional: Create a namespace for the issuer.

.. Create a YAML file that defines a `Namespace` object:
+
.Example `namespace.yaml` file
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: my-issuer-namespace <1>
----
<1> Specify the namespace for the issuer.

.. Create the `Namespace` object by running the following command:
+
[source,terminal]
----
$ oc create -f namespace.yaml
----

> 

Reviewer 2: Here there is another additional step needed for creating the secret itself (similar to what you did in AWS explicit).

eg. 
```
$ oc create secret generic clouddns-dns01-solver-svc-acct --from-file=service_account.json=<path/to/.docker/gcp_service_account.json> -n my-issuer-namespace
```

. Create an issuer.

.. Create a YAML file that defines the `Issuer` object:
+
.Example `issuer.yaml` file
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: issuer-acme-dns01-clouddns <1>
spec:
  acme:
    preferredChain: ""
    privateKeySecretRef:
      name: <secret_private_key> <2>
    server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
    solvers:
    - dns01:
        cloudDNS:
          project: <project_id> <4>
          serviceAccountSecretRef:
            name: clouddns-dns01-solver-svc-acct <5>

>             key: key.json <6>

Reviewer 2: ```suggestion
            key: service_account.json <6>
```
Let's make it `service_account.json` as that is the default name (and also the name that I suggested in previous step).

----
<1> Provide a name for the issuer.
<2> Replace `<secret_private_key>` with the name of the secret to store the ACME account private key in.
<3> Specify the URL to access the ACME server's `directory` endpoint. This example uses the _Let's Encrypt_ staging environment.

> <4> Replace `<project_id>` with the name of the project.

Reviewer 2: `name of the GCP project which contains the Cloud DNS zone`

just being specific that it's about the GCP project id.

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers-only'       <2>
----
<1> Add the `spec.controllerConfig` section.
<2> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.

.. Save the file to apply the changes.

. Optional: Create a namespace for the issuer.

.. Create a YAML file that defines a `Namespace` object:
+
.Example `namespace.yaml` file
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: my-issuer-namespace <1>
----
<1> Specify the namespace for the issuer.

.. Create the `Namespace` object by running the following command:
+
[source,terminal]
----
$ oc create -f namespace.yaml
----

. Create a secret to store your GCP credentials by running the following command:
+
[source,terminal]
----

> $ oc create secret generic clouddns-dns01-solver-svc-acct --from-file=service_account.json=<path/to/.docker/gcp_service_account.json> -n my-issuer-namespace

Reviewer 2: My absolute bad, `.docker/` isn't required.
Please make it `--from-file=service_account.json=<path/to/gcp_service_account.json>` instead.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-explicit-gcp_{context}"]
= Configuring an ACME issuer to solve DNS-01 challenges by using explicit credentials for GCP Cloud DNS

You can use the {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges by using explicit credentials on GCP. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Google CloudDNS.

.Prerequisites

* You have access to a role that has a permission to create issuers, certificates, and secrets in the desired namespace.
* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
* You have set up an IAM role for Google CloudDNS. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/google/[Google CloudDNS] in the upstream cert-manager documentation.

> 

Reviewer 2: Please add a note similar to AWS explicit:
```
You can use Google Cloud DNS with explicit credentials in an OpenShift Container Platform cluster that is not running on GCP.
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-explicit-gcp_{context}"]
= Configuring an ACME issuer to solve DNS-01 challenges by using explicit credentials for GCP Cloud DNS

You can use the {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges by using explicit credentials on GCP. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Google CloudDNS.

.Prerequisites

> * You have access to a role that permissions to create issuers, certificates, and secrets in the desired namespace.

Reviewer 2: ```suggestion
* You have access to a role that has permissions to create issuers, certificates, and secrets in the desired namespace.
```

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.

> * You have set up an IAM role for Google CloudDNS. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/google/[Google CloudDNS] in the upstream cert-manager documentation.

Reviewer 3: > You have set up an IAM role for Google CloudDNS

Incorrect. For GCP, it is not only IAM role, it is Google Cloud service account of right roles.

Reviewer 2: Doesn't GCP still use "IAM" keyword in it's service when referring to service accounts? I'd prefer "service account" though.

Reviewer 3: > For GCP, it is not **only** IAM role

I mean it is not **ONLY** IAM role. So I suggest "Google Cloud service account of right roles"

+
[NOTE]
====
You can use Google CloudDNS with explicit credentials in an {product-title} cluster that is not running on GCP.
====

.Procedure

. Optional: Override the nameserver settings for the DNS-01 self check:
+
This step is required only when the target public-hosted zone overlaps with the cluster's default private-hosted zone.

.. Edit the `CertManager` resource by running the following command:
+
[source,terminal]
----
$ oc edit certmanager cluster
----

.. Add a `spec.controllerConfig` section with the following override arguments:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
  ...
spec:
  ...
  controllerConfig:                                <1>
    overrideArgs:
      - '--dns01-recursive-nameservers-only'       <2>
----
<1> Add the `spec.controllerConfig` section.
<2> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.

> 

Reviewer 3: This is incorrect. There are 3 occurrences where it only documents --dns01-recursive-nameservers-only without --dns01-recursive-nameservers=1.1.1.1:53. All occurrences should document both.

Reviewer 3: But this is not a suggested documenting way. Reading https://59702--docspreview.netlify.app/openshift-enterprise/latest/security/cert_manager_operator/cert-manager-operator-issuer-acme.html#cert-manager-acme-dns01-explicit-azure_cert-manager-operator-issuer-acme , it shows 5 occurrences of `Optional: Override the nameserver ...`, it is redundant, duplicate and easy to be inconsistent as pointed out.
Suggest please not duplicate and just put this in one occurrence somewhere for example at the beginning of all dns-01 sections.

Author: @xingxingxia So considering this structure, right now we cannot avoid those redundant instructions. Because for eg, if a customer jumps on any particular section/module he will follow the sequential steps without referring to any other section. So we must provide step-by-step instructions. To avoid duplicate steps I can think of restructuring but that might take some time. We can get this structure merged and then I can work on restructuring OR if we can wait then I can start working on restructuring immediately.

Reviewer 2: I'd suggest that for the time being we keep the current structure, cause though it contains some redundant information it is informative for users with completeness following particular sub sections to achieve what they're looking for through the steps in the docs exclusively, as long as we have correctness of the functionality. IMO, we probably could focus on getting this out and later we can revisit to remove some of the redundancies through plausible restructuring.

Reviewer 3: OK, it is fine, but please ensure `All occurrences should document both`. Now there is still one occurrence under `ambient credentials on GCP` section which still only documents --dns01-recursive-nameservers-only without --dns01-recursive-nameservers=1.1.1.1:53 .

Reviewer 3: Copying and pasting would work, but one nagging here, not sure why some some occurrences of same snippets are:
```
      - '--dns01-recursive-nameservers-only'       
      - '--dns01-recursive-nameservers=1.1.1.1:53'
```
While other occurrences are inconsistent order:
```
      - '--dns01-recursive-nameservers=1.1.1.1:53' 
      - '--dns01-recursive-nameservers-only'  
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc

:_content-type: PROCEDURE
[id="cert-manager-acme-dns01-explicit-gcp_{context}"]
= Configuring an ACME issuer to solve DNS-01 challenges by using explicit credentials for GCP Cloud DNS

You can use the {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges by using explicit credentials on GCP. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Google CloudDNS.

.Prerequisites

> * You have access to the {product-title} cluster as a user with the `cluster-admin` role.

Reviewer 3: Not sure for which steps this prerequisite is. In this section, `oc edit certmanager cluster` needs it, `oc edit certmanager cluster` appears in all other dns-01 sections, but those sections do not list this prerequisite, this is inconsistent.

- If this prerequisite is is for `oc edit certmanager cluster`, then all other dns-01 sections need add it.

Reviewer 3: But if this prerequisite is not extended for `oc edit certmanager cluster`, then let's look at remaining snippets:
oc create -f namespace.yaml
oc create -f issuer.yaml
oc create -f certificate.yaml

Only namespace creation `oc create -f namespace.yaml` needs cluster-admin, Issuer and Certificate creation need not. So, if this prerequisite is added for namespace creation only, I suggest replace it with `oc new-project my-issuer-namespace`, `oc new-project` needs not cluster-admin, as previous comment https://github.com/openshift/openshift-docs/pull/59702#discussion_r1233798844 .

Author: Added `oc new-project my-issuer-namespace` in all relevant places
